### PR TITLE
DenseMap: support enum class keys

### PR DESCRIPTION
**Pull request by artagnon**
Implemented using std::underlying_type.

### CODE REVIEW

File: llvm/include/llvm/ADT/DenseMapInfo.h [hunk 1]
```diff
@@ -297,6 +297,27 @@ template <typename... Ts> struct DenseMapInfo<std::tuple<Ts...>> {
...
+    return static_cast<Enum>(Info::getTombstoneKey());
+  }
+
+  static unsigned getHashValue(const Enum &Val) {
+    return Info::getHashValue(static_cast<UnderlyingType>(Val));
+  }
+
+  static bool isEqual(const Enum &LHS, const Enum &RHS) {
+    return Info::isEqual(static_cast<UnderlyingType>(LHS),
+                         static_cast<UnderlyingType>(RHS));
```
Comment (nikic) [inline llvm/include/llvm/ADT/DenseMapInfo.h:318]: Could directly compare `LHS == RHS` here?

File: llvm/unittests/ADT/MapVectorTest.cpp [hunk 2]
```diff
@@ -267,6 +267,30 @@ TEST(MapVectorTest, NonCopyable) {
   ASSERT_EQ(*MV.find(2)->second, 2);
 }
 
+TEST(MapVectorTest, EnumClassKey) {
```
Comment (nikic) [inline llvm/unittests/ADT/MapVectorTest.cpp:270]: Why test this through MapVector rather than DenseMap directly?

**Review by nikic — COMMENTED** [review]
Looks reasonable.

**Comment by nikic** [comment]
Though I would like to hear where you plan to use this. Usually enums imply that there is a reasonably small, continuously numbered and fixed set of keys, in which case using a vector instead of map is preferable.

File: llvm/unittests/ADT/MapVectorTest.cpp (hunk 2 above)
Comment (artagnon) [reply llvm/unittests/ADT/MapVectorTest.cpp:270]: There's no DenseMapTest.cpp. Perhaps just use DenseMap directly in MapVectorTest.cpp?

Comment (nikic) [reply llvm/unittests/ADT/MapVectorTest.cpp:270]: I see a DenseMapTest.cpp here: https://github.com/llvm/llvm-project/blob/main/llvm/unittests/ADT/DenseMapTest.cpp

**Comment by artagnon** [comment]
> Though I would like to hear where you plan to use this. Usually enums imply that there is a reasonably small, continuously numbered and fixed set of keys, in which case using a vector instead of map is often preferable.

This is a requirement for my new Presburger parser, which keeps a map of `{DimOrSymbol, position} -> multiplier`. The parser isn't ready just yet, but will be soon'ish.

**Comment by nikic** [comment]
> > Though I would like to hear where you plan to use this. Usually enums imply that there is a reasonably small, continuously numbered and fixed set of keys, in which case using a vector instead of map is often preferable.
> 
> This is a requirement for my new Presburger parser, which keeps a map of `{DimOrSymbol, position} -> multiplier`. The parser isn't ready just yet, but will be soon'ish.

Ah, so this is mostly a building block for hashing std::pair? That makes sense.

**Review by kuhar — APPROVED** [review]

**Pull request merged**